### ISSUE TITLE
How does one use -J and compile_commands.json?

### ISSUE THREAD

**Issue by BillWhite** [Andersbakken/rtags#252]
I'm trying to use rtags on rtags.  Presumbly that works, somehow.  I used ninja to
generate compile_commands.json and put it in the build area, not the source area.
I then ran rdm with working directory in the build area.  I put .rtags-config at the
root of the source directory, with the project: iine set up correctly (I think).  I can see
that rc commands are sent to rdm, but rdm does not ever seem to index anything.

I would have thought that running "rc -J $buildArea" would send all the files in
$buildArea/compile_commands.json to rdm for processing.  But apparently I misunderstand.

Thanks.

**Comment by YorkZ** [comment]
Hi Andersbakken,

I've sent the compile_commands.json file to you. Please check your mailbox.

Thanks,
York

**Comment by Andersbakken** [comment]
Hi York

It seems like there's nothing but preprocessing commands in the json file.
I'm a little confused as to how anything gets compiled here.

Anders

**Comment by YorkZ** [comment]
Hi Andersbakken,

Thank you very much for the help.

> I'm a little confused as to how anything gets compiled here

I generated the compile_commands.json file using Bear. Looks like that bear
didn't do the correct job. Is there any other tools you can recommend me to try
to generate a JSON compilation database?

York

**Comment by Andersbakken** [comment]
I've always used cmake's built-in feature for this but if you're not using
cmake that's not much help. To be honest I kinda don't rely on
compile_commands.json but rather putting rtags-gcc-wrapper.sh in my path
symlinked as c++/cc/gcc/g++ etc.

Anders

**Comment by YorkZ** [comment]
Yes, cmake's built-in feature to generate the compilation database is really handy. But unfortunately our project doesn't use cmake. I've used `rtags-gcc-wrapper.sh` to intercept the compilation command and it was nice too except that I have these problems:

1. Instead of issuing `rc -C` to clear everything in ~/.rtags, is there a way to clear the data only for one project without touching the index files of other projects?
2. Is there a way to re-index a project from scratch without having to re-build the project? This is extremely useful for huge projects which could take a few hours to build, and is absolutely indispensable in my case here when I was trying to find out which line of the code was causing libclang to crash.
3. How do I pass additional options to libclang without having to change the `Makefile`? Let's say I want to pass `-Wno-mismatched-tags` to libclang, if I use a compilation database, I could modify the file `compile_commands.json` to add this option. Without a compilation database, how do I achieve this? On the other hand, even if I have a compilation database, being able to specify additional options is useful because one wouldn't have to modify the compilation database by hand, especially if the project contains thousands of files.
4. Similar to 3, is there a way to delete an option without having to change the `Makefile`? This and the previous one are extremely useful because in my case I need to pass certain options only to libclang while indexing the project, but not to GCC while compiling the project.
5. Now that we can use `rtags-gcc-wrapper.sh` to intercept GCC command line, is it feasible or how hard is it for `rtags` to generate a compilation database?

I'm sorry for having asked so many questions.

Thanks again,
York

**Comment by Andersbakken** [comment]
Hi

On Sunday, January 4, 2015, York Zhao <notifications@github.com> wrote:

> Yes, cmake's built-in feature to generate the compilation database is
> really handy. But unfortunately our project doesn't use cmake. I've used
> rtags-gcc-wrapper.sh to intercept the compilation command and it was nice
> too except that I have these problems:
>
>    1. Instead of issuing rc -C to clear everything in ~/.rtags, is there
>    a way to clear the data only for one project without touching the index
>    files of other projects?
>
>
rc -W [pattern that matches the project you want to remove]

>
>    1. Is there a way to re-index a project from scratch without having to
>    re-build the project? This is extremely useful for huge projects which
>    could take a few hours to build, and is absolutely indispensable in my case
>    here when I was trying to find out which line of the code was causing
>    libclang to crash.
>
>
rc -V [pattern that matches the files you want to dirty or empty to dirty
everything]

>
>    1. How do I pass additional options to libclang without having to
>    change the Makefile? Let's say I want to pass -Wno-mismatched-tags to
>    libclang, if I use a compilation database, I could modify the file
>    compile_commands.json to add this option. Without a compilation
>    database, how do I achieve this? On the other hand, even if I have a
>    compilation database, being able to specify additional options is useful
>    because one wouldn't have to modify the compilation database by hand,
>    especially if the project contains thousands of files.
>
>

>    1. Similar to 3, is there a way to delete an option without having to
>    change the Makefile? This and the previous one are extremely useful
>    because in my case I need to pass certain options only to libclang while
>    indexing the project, but not to GCC while compiling the project.
>
>
I just added something. You can now do:

./rdm --arg-transform ~/scripts/rtags-arg-transform.sh

And then you can modify the arguments in the script or even return non-0 to
cancel the whole compile command.

E.g.

# cat ~/scripts/rtags-arg-transform.sh

#!/bin/bash

echo $@ -Wno-mismatched-tags | sed -e 's,-Wfoobar,,'


>    1. Now that we can use rtags-gcc-wrapper.sh to intercept GCC command
>    line, is it feasible or how hard is it for rtags to generate a
>    compilation database?
>
> RTags can't currently do it but I'll look into adding something for it. I
think it's a good idea.


>
>
> I'm sorry for having asked so many questions.
>

No problem at all. I'm happy people are using RTags.

> Thanks again,
> York
>
My pleasure.

Anders

**Comment by YorkZ** [comment]
Hi Anders,

> rc -W [pattern that matches the project you want to remove]

I was aware `rc -W` and I had actually tried it. However, my problem was that my projects are like:

/home/york/foo
/home/york/foo_bar

I wanted to delete index for project `foo` but not `foo_bar`. I've tested `rc -w foo$` to list only the project `foo` but it didn't work. Am I missing anything here? Also, I think it would be safer to have a "dry run" mechanism to allow the user to test his matching regexp before actually applying it to delete anything. Is it OK to use `-w` for this purpose like what I did?

> rc -V [pattern that matches the files you want to dirty or empty to dirty everything]

I have the same problem as with `-W`

`rdm --arg-transform` looks like a cool idea, I will try it out. Thank you so much for quickly adding this feature.

I'm glad that you agree with my idea of generating a JSON compilation database. I look forward to trying out this feature once you complete it.

Thank you very much,
York

**Comment by YorkZ** [comment]
> I've tested rc -w foo$ to list only the project foo but it didn't work.

Forget about it. I just tried `rc -w foo/$` instead of `rc -w foo$` and it worked. However, I just realized that `rc -w foo/$` would select the project `foo`, not listing the project. Therefore, I guess having a "dry run" feature for `rc -W` may still be desirable?

Thanks,
York

**Comment by Andersbakken** [comment]
Couldn't this be achieved with shell tools?

What would you have the command do?

E.g.

rc -w

/src/foobar/
/src/foobar2/
/src/foobar3/

rc -W foobar --dry-run

Deleted project: /src/foobar/
Deleted project: /src/foobar2/
Deleted project: /src/foobar3/

But in the end it didn't happen?

Anders

**Comment by Andersbakken** [comment]
Btw. I added the compile_commands.json feature. You can now do:

rc --dump-compilation-database

Anders

**Comment by YorkZ** [comment]
> Btw. I added the compile_commands.json feature

Just tested this on a "Hello world" project and seems to work great. I'll test it later today on a real project where [bear fails to generate a compilation database.](https://github.com/rizsotto/Bear/issues/69). Thank you so much for this useful feature. It's especially useful for projects where "bear" fails!

**Comment by YorkZ** [comment]
E.g.

> rc -w
> /src/foobar/
> /src/foobar2/
> /src/foobar3/
>
> ...
>
> rc -W foobar --dry-run
>
> Deleted project: /src/foobar/
> Deleted project: /src/foobar2/
> Deleted project: /src/foobar3/

> But in the end it didn't happen?

Yes. Or, maybe it's easier to print the name of the projects that will be deleted and ask for a "y/n" confirmation from the user?

**Comment by YorkZ** [comment]
Hi Andersbakken,

I've tested the "--dump-compilation-database" feature against the project where [bear fails to generate a compilation database](https://github.com/rizsotto/Bear/issues/69), and it seemed to work.

By the way, I've used `rtags` during the last week on a huge project. It worked way too good and has helped me a lot. For C++ source code, this is 100 times better than any other simple tag system. I really really really love it, and thank you very much for this super useful tool.

The only thing that is not so nice is the Emacs `rtags-mode` buffer. The readability is poor when the file paths are long. I'm going to improve it and will send you a pull request when I complete it.
